File: libraries/python/tests/manual_stress_manager.py
"""Manual stress test: fast and slow simulations without plots, with tracking.

Run from ``libraries/python`` with::

    uv run python -m tests.manual_stress_manager

Requires TRNSYS, Type3830, and the bundled TRNRun/queue executables. The slow
fixture also requires its referenced TRNSYS weather file. Edit the configuration
below for your installation. Deck copies and outputs are retained directly in
``tests/runs`` for inspection. This module is deliberately not named ``test_*``
so pytest will not collect it.
"""

from __future__ import annotations

import shutil
import sys
from pathlib import Path
from time import perf_counter

from trnrun import SimulationConfig, SimulationManager

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
__root__ = Path(__file__).resolve().parent

TRNEXE_PATH = Path(r"C:\TRNSYS18\Exe\TrnEXE64.exe")
FAST_DCK = __root__ / "dck" / "test_fast_wo_plot_w_tracking.dck"
SLOW_DCK = __root__ / "dck" / "test_slow_wo_plot_w_tracking.dck"
DCK_FOLDER = __root__ / "runs"

FAST_SIM_COUNT = 0
SLOW_SIM_COUNT = 50
MAX_CONCURRENT = 25
REFRESH_INTERVAL = 0.1

CONFIG = SimulationConfig(
    trnexe_path=TRNEXE_PATH,
    watch_tmp=True,
)


# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------
def copy_dck(src: Path, dst_dir: Path, n: int) -> list[Path]:
    """Copy a fixture with unique, zero-padded names for independent outputs."""
    dst_files: list[Path] = []

    for i in range(1, n + 1):
        dst = dst_dir / f"{src.stem}_{i:04d}{src.suffix}"
        _ = shutil.copyfile(src, dst)
        dst_files.append(dst)

    return dst_files


def run_simulations(dck_files: list[Path]) -> SimulationManager:
    """Submit both workloads to one manager and wait for all runs to finish."""
    with SimulationManager(
        max_concurrent=MAX_CONCURRENT,
        refresh_interval=REFRESH_INTERVAL,
    ) as manager:
        for dck in dck_files:
            _ = manager.add(dck, CONFIG)

        manager.wait()

    return manager


# -----------------------------------------------------------------------------
# Entry point
# -----------------------------------------------------------------------------
def main() -> int:
    """Prepare the stress workload and return a failing exit code unless all succeed."""
    CONFIG.validate()

    for source in (FAST_DCK, SLOW_DCK):
        if not source.is_file():
            raise FileNotFoundError(f"Deck file not found: {source}")

    DCK_FOLDER.mkdir(parents=True, exist_ok=True)

    # print(f"Decks and outputs: {DCK_FOLDER}", flush=True)

    dck_files = copy_dck(FAST_DCK, DCK_FOLDER, FAST_SIM_COUNT)
    dck_files.extend(copy_dck(SLOW_DCK, DCK_FOLDER, SLOW_SIM_COUNT))

    print(f"Running {FAST_SIM_COUNT:,} fast + {SLOW_SIM_COUNT:,} slow simulations (concurrency: {MAX_CONCURRENT}).")

    started = perf_counter()
    manager = run_simulations(dck_files)
    elapsed = perf_counter() - started

    print(
        f"Finished in {elapsed:.1f}s: ",
        f"{len(manager.succeeded):,}/{len(dck_files):,} succeeded, ",
        f"{len(manager.failed):,} failed.",
    )

    for simulation in manager.failed:
        print(
            f"FAILED {simulation.deck_path.name}: status={simulation.status}, completion={simulation.completion_event}",
        )

    return 0 if len(manager.succeeded) == len(dck_files) else 1


if __name__ == "__main__":
    sys.exit(main())

File: libraries/matlab/tests/manual_stress_manager.m
function exit_code = manual_stress_manager()
%MANUAL_STRESS_MANAGER Stress-test fast and slow tracked simulations.
%   EXIT_CODE = MANUAL_STRESS_MANAGER() copies fast and slow deck fixtures,
%   submits them through one TRNRun manager, and returns zero only when every
%   simulation succeeds.
%
%   Run from libraries/matlab with:
%
%       matlab -batch "addpath('tests'); exit(manual_stress_manager())"
%
%   This test requires TRNSYS, Type3830, and the bundled TRNRun and queue
%   executables. The slow fixture also requires its referenced TRNSYS weather
%   file. Edit the configuration below for your installation. Deck copies and
%   outputs are retained directly in tests/runs for inspection.

    tests_folder = string(fileparts(mfilename("fullpath")));
    matlab_folder = fileparts(tests_folder);
    addpath(fullfile(matlab_folder, "toolbox"));
    import trnrun.*

    % -------------------------------------------------------------------------
    % Configuration
    % -------------------------------------------------------------------------
    trnexe_path = "C:\TRNSYS18\Exe\TrnEXE64.exe";
    fast_dck = fullfile(tests_folder, "dck", ...
        "test_fast_wo_plot_w_tracking.dck");
    slow_dck = fullfile(tests_folder, "dck", ...
        "test_slow_wo_plot_w_tracking.dck");
    dck_folder = fullfile(tests_folder, "runs");

    fast_sim_count = 1;
    slow_sim_count = 20;
    max_concurrent = 10;
    refresh_interval = 0.1;

    config = SimulationConfig( ...
        trnexe_path=trnexe_path, ...
        watch_tmp=true);

    % -------------------------------------------------------------------------
    % Prepare workload
    % -------------------------------------------------------------------------
    config = config.validate();

    sources = [fast_dck, slow_dck];
    for source = sources
        if ~isfile(source)
            error("trnrun:DeckNotFound", "Deck file not found: %s", source);
        end
    end

    if ~isfolder(dck_folder)
        mkdir(dck_folder);
    end

    fprintf("Decks and outputs: %s\n", dck_folder);

    dck_files = copy_dck(fast_dck, dck_folder, fast_sim_count);
    dck_files = [dck_files, ...
        copy_dck(slow_dck, dck_folder, slow_sim_count)];

    fprintf("Running %d fast + %d slow simulations (concurrency: %d).\n", ...
        fast_sim_count, slow_sim_count, max_concurrent);

    % -------------------------------------------------------------------------
    % Run
    % -------------------------------------------------------------------------
    started = tic;
    manager = SimulationManager( ...
        maxConcurrent=max_concurrent, ...
        refreshInterval=refresh_interval);

    for dck = dck_files
        manager.add(dck, config);
    end

    manager.wait();
    manager.shutdown();
    elapsed = toc(started);

    succeeded = manager.succeeded;
    failed = manager.failed;

    fprintf("Finished in %.1fs: %d/%d succeeded, %d failed.\n", ...
        elapsed, numel(succeeded), numel(dck_files), numel(failed));

    for simulation = failed
        fprintf("FAILED %s: status=%s, completion=%s\n", ...
            deck_name(simulation.deckPath), ...
            event_json(simulation.status), ...
            event_json(simulation.completionEvent));
    end

    exit_code = double(numel(succeeded) ~= numel(dck_files));
end

function dck_files = copy_dck(src, dst_dir, n)
%COPY_DCK Copy SRC N times with unique, zero-padded names.

    [~, stem, extension] = fileparts(src);
    dck_files = strings(1, n);

    for index = 1:n
        destination = fullfile(dst_dir, ...
            sprintf("%s_%04d%s", stem, index, extension));
        copyfile(src, destination, "f");
        dck_files(index) = destination;
    end
end

function name = deck_name(path)
%DECK_NAME Return the filename portion of a deck path.

    [~, stem, extension] = fileparts(path);
    name = stem + extension;
end

function text = event_json(event)
%EVENT_JSON Render an event struct for failure diagnostics.

    if isempty(event)
        text = "[]";
    else
        text = string(jsonencode(event));
    end
end
